base/opaque_closure.jl

```diff
@@ -110,3 +110,197 @@ function generate_opaque_closure(@nospecialize(sig), @nospecialize(rt_lb), @nosp
     return ccall(:jl_new_opaque_closure_from_code_info, Any, (Any, Any, Any, Any, Any, Cint, Any, Cint, Cint, Any, Cint, Cint),
         sig, rt_lb, rt_ub, mod, src, lineno, file, nargs, isva, env, do_compile, isinferred)
 end
+
+struct Slot{T} end
+struct Splat{T}
+    value::T
+end
+
+# Args is a Tuple{Vararg{Union{Slot{T},Some{T}}}} where Slot{T} represents
+# an uncurried argument slot, and Some{T} represents an argument to curry.
+@noinline @generated function Core.OpaqueClosure(Args::Tuple, ::Slot{RT}) where RT
```

Comment:
This seems unnecessarily complicated. Why not have `AT` be passthrough and specify `(nreq, isva)` as a Val?

Reply:
The idea was to support syntaxes like `@TypedCallable add_node(self, ::Node)` or `@TypedCallable show(self)::Nothing` where we close over more than just the first argument

Reply:
It is also banned in precompile, since `@generated` is not permitted to return a new `:opaque_closure` object. Can we do without this hand-generated complexity? Stuff like `Args.parameters` is typically not actually recommended in a generated function either, as it returns something with incorrect type identity (makes the transofrm not pure). I remember doing something like `make(f, AT, RT) = (Base.inferencebarrier(Base.Experimental.@opaque(AT->RT, (args...)->f(args...)))::Core.OpaqueClosure)
`

The Slot/Splat seems to be just a partial re-implementation of lambdas, but seems a bit less reliable since it has none of the normal lowering, and makes it so that the call is not a subtype of its argument signature?

Reply:
> It is also banned in precompile, since @generated is not permitted to return a new :opaque_closure object.

this opts-out of PartialOpaque support via https://github.com/JuliaLang/julia/pull/54734 so that this is allowed in pre-compile

Reply:
It is not inference that is banned, it is the construct itself, since it allocates new state (a Method) which is forbidden during pure operations (a generator)

Reply:
Ah I see - I thought that was only forbidden during the execution of the generator, but guess it also applies to the side effects of lowering the generated expression?